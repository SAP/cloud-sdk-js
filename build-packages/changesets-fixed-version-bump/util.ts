/* eslint-disable jsdoc/require-jsdoc */

import { PathLike } from 'node:fs';
import { readFile } from 'node:fs/promises';
import { info } from 'node:console';
import getReleasePlan from '@changesets/get-release-plan';
import { inc } from 'semver';

export async function getPackageVersion(
  pathToRootPackageJson?: PathLike
): Promise<string> {
  const packageJson = await readFile(
    pathToRootPackageJson || 'package.json',
    'utf8'
  );
  return JSON.parse(packageJson).version;
}

const bumpTypeOrder = ['major', 'minor', 'patch', 'none'] as const;

export async function getNextVersion(): Promise<{
  version: string;
  bumpType: (typeof bumpTypeOrder)[number];
}> {
  const currentVersion = await getPackageVersion();
  info(`Current version: ${currentVersion}`);

  const bumpType = await getBumpType();
  info(`Bump type: ${bumpType}`);

  if (bumpType === 'none' || !bumpType) {
    throw new Error(`No changesets to release`);
  }

  const version = inc(currentVersion, bumpType);

  if (!version) {
    throw new Error(
      `Invalid new version -- current version: ${currentVersion}, bump type: ${bumpType}`
    );
  }
  return { version, bumpType };
}

async function getBumpType() {
  const releasePlan = await getReleasePlan(process.cwd());
  info(`Release plan: ${JSON.stringify(releasePlan)}`);

  const versionIncreases = releasePlan.releases
    .map(({ type }) => bumpTypeOrder.indexOf(type))
    .sort((a, b) => b - a);
  return bumpTypeOrder[Math.min(...versionIncreases)];
}
