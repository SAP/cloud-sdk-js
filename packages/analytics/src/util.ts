/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import crypto from 'crypto';

/**
 * Generate a random salt.
 * @returns A random salt.
 * @hidden
 */
export function randomSalt(): string {
  // 1 byte will produce two bytes in hex string
  // 32 random bytes generate a hex string of the same bitlength as SHA-256
  return crypto.randomBytes(32).toString('hex');
}

/**
 * Create a hash for the given value.
 * @param value - The value to hash.
 * @returns The hashed value.
 * @hidden
 */
export function hash(value: string): string {
  return crypto
    .createHash('sha256')
    .update(value)
    .digest('hex');
}
