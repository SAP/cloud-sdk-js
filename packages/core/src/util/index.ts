export * from './jwt';
export * from './name-converter';
export * from './properties-util';
export * from './proxy-util';
export * from './remove-slashes';
