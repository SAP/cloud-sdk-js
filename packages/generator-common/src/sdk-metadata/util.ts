import { resolve } from 'path';
import { readFile } from 'fs-extra';

/**
 * Get the current SDK version from the package json.
 * @returns The SDK version.
 * @internal
 */
export async function getSdkVersion(): Promise<string> {
  return JSON.parse(
    await readFile(resolve(__dirname, '../../package.json'), 'utf8')
  ).version;
}
