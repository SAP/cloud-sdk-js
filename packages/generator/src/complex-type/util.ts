import { VdmProperty } from '../vdm-types';

/**
 * @internal
 */
export function hasEdmTypeProperty(properties: VdmProperty[]): boolean {
  return properties.some(prop => !prop.isComplex && !prop.isEnum);
}
/**
 * @internal
 */
export function hasComplexTypeProperty(properties: VdmProperty[]): boolean {
  return properties.find(prop => prop.isComplex === true) !== undefined;
}
