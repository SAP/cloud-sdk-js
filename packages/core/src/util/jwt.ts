/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { IncomingMessage } from 'http';
import { createLogger, errorWithCause } from '@sap-cloud-sdk/util';
import { AxiosRequestConfig } from 'axios';
import jwt from 'jsonwebtoken';
import {
  Cache,
  fetchVerificationKeys,
  getXsuaaServiceCredentials,
  TokenKey,
  XsuaaServiceCredentials
} from '../scp-cf';
import {
  mapping as mappingTenantFields,
  RegisteredJWTClaimsTenant
} from '../scp-cf/tenant';
import {
  mapping as mappingUserFields,
  RegisteredJWTClaimsUser,
  Scope
} from '../scp-cf/user';

const logger = createLogger({
  package: 'core',
  messageContext: 'jwt'
});

/**
 * Decode JWT.
 * @param token - JWT to be decoded
 * @returns Decoded payload.
 */
export function decodeJwt(token: string): DecodedJWT {
  return decodeJwtComplete(token).payload;
}

/**
 * Decode JWT and return the complete decoded token.
 * @param token - JWT to be decoded
 * @returns Decoded token containing payload, header and signature.
 */
export function decodeJwtComplete(token: string): CompleteDecodedJWT {
  const decodedToken = jwt.decode(token, { complete: true });
  if (decodedToken === null || typeof decodedToken === 'string') {
    throw new Error(
      'JwtError: The given jwt payload does not encode valid JSON.'
    );
  }
  return {
    header: decodedToken.header,
    payload: decodedToken.payload,
    signature: decodedToken.signature
  };
}

/**
 * Retrieve JWT from a request that is based on the node `IncomingMessage`. Fails if no authorization header is given or has the wrong format. Expected format is 'Bearer <TOKEN>'.
 * @param req - Request to retrieve the JWT from
 * @returns JWT found in header
 */
export function retrieveJwt(req: IncomingMessage): string | undefined {
  const header = authHeader(req);
  if (validateAuthHeader(header)) {
    return header!.split(' ')[1];
  }
}

function authHeader(req: IncomingMessage): string | undefined {
  const entries = Object.entries(req.headers).find(
    ([key]) => key.toLowerCase() === 'authorization'
  );
  if (entries) {
    const header = entries[1];

    // Header could be a list of headers
    return Array.isArray(header) ? header[0] : header;
  }
  return undefined;
}

function validateAuthHeader(header: string | undefined): boolean {
  if (typeof header === 'undefined') {
    logger.warn('Authorization header not set.');
    return false;
  }

  const [authType, token] = header.split(' ');

  if (typeof token === 'undefined') {
    logger.warn('Token in auth header missing.');
    return false;
  }

  if (authType.toLowerCase() !== 'bearer') {
    logger.warn('Authorization type is not Bearer.');
    return false;
  }

  return true;
}

/**
 * Verifies the given JWT and returns the decoded payload.
 *
 * @param token - JWT to be verified
 * @param options - Options to control certain aspects of JWT verification behavior.
 * @returns A Promise to the decoded and verified JWT.
 */
export async function verifyJwt(
  token: string,
  options?: VerifyJwtOptions
): Promise<DecodedJWT> {
  options = { ...defaultVerifyJwtOptions, ...options };

  const creds = getXsuaaServiceCredentials(token);
  const verificationKeyURL = getVerificationKeyURL(token);

  if (
    options.cacheVerificationKeys &&
    verificationKeyCache.get(verificationKeyURL)
  ) {
    const key = verificationKeyCache.get(verificationKeyURL) as TokenKey;

    return verifyJwtWithKey(token, key.value).catch(error => {
      logger.warn(
        'Unable to verify JWT with cached key, fetching new verification key.'
      );
      logger.warn(`Original error: ${error.message}`);

      return fetchAndCacheKeyAndVerify(
        creds,
        verificationKeyURL,
        token,
        options
      );
    });
  }

  return fetchAndCacheKeyAndVerify(creds, verificationKeyURL, token, options); // Verify only here
}

function fetchAndCacheKeyAndVerify(
  creds: XsuaaServiceCredentials,
  verificationKeyURL: string,
  token: string,
  options?: VerifyJwtOptions
) {
  return getVerificationKey(creds, verificationKeyURL)
    .catch(error =>
      Promise.reject(
        errorWithCause(
          'Failed to verify JWT - unable to get verification key!',
          error
        )
      )
    )
    .then(key =>
      options ? cacheVerificationKey(verificationKeyURL, key, options) : key
    )
    .then(key => verifyJwtWithKey(token, key.value));
}

/**
 * Options to control certain aspects of JWT verification behavior.
 */
export interface VerifyJwtOptions {
  cacheVerificationKeys?: boolean;
}

const defaultVerifyJwtOptions: VerifyJwtOptions = {
  cacheVerificationKeys: true
};

function getVerificationKey(
  xsuaaCredentials: XsuaaServiceCredentials,
  jku: string
): Promise<TokenKey> {
  return fetchVerificationKeys(xsuaaCredentials, jku).then(verificationKeys => {
    if (!verificationKeys.length) {
      throw Error(
        'No verification keys have been returned by the XSUAA service!'
      );
    }
    return verificationKeys[0];
  });
}

// 15 minutes is the default value used by the xssec lib
export const verificationKeyCache = new Cache({ minutes: 15 });

function cacheVerificationKey(
  verificationKeyURL: string,
  key: TokenKey,
  options: VerifyJwtOptions
): TokenKey {
  if (options.cacheVerificationKeys) {
    verificationKeyCache.set(verificationKeyURL, key);
  }
  return key;
}

/**
 * Verifies the given JWT with the given key and returns the decoded payload.
 *
 * @param token - JWT to be verified
 * @param key - Key to use for verification
 * @returns A Promise to the decoded and verified JWT.
 */
export function verifyJwtWithKey(
  token: string,
  key: string
): Promise<DecodedJWT> {
  return new Promise((resolve, reject) => {
    jwt.verify(token, sanitizeVerificationKey(key), (err, decodedToken) => {
      if (err) {
        reject(errorWithCause('JWT invalid', err));
      } else {
        resolve(decodedToken as DecodedJWT);
      }
    });
  });
}

function sanitizeVerificationKey(key: string) {
  // Add new line after -----BEGIN PUBLIC KEY----- and before -----END PUBLIC KEY----- because the lib won't work otherwise
  return key
    .replace(/\n/g, '')
    .replace(/(KEY\s*-+)([^\n-])/, '$1\n$2')
    .replace(/([^\n-])(-+\s*END)/, '$1\n$2');
}

/**
 * Compare two decoded JWTs based on their tenantIds.
 * @param decodedUserToken - User JWT
 * @param decodedProviderToken - Provider JWT
 * @returns Whether the tenant is identical.
 */
export function isIdenticalTenant(
  decodedUserToken: DecodedJWT,
  decodedProviderToken: DecodedJWT
): boolean {
  return (
    readPropertyWithWarn(decodedUserToken, mappingTenantFields.id.keyInJwt) ===
    readPropertyWithWarn(decodedProviderToken, mappingTenantFields.id.keyInJwt)
  );
}

/**
 * Get the issuer url of a decoded JWT.
 * @param decodedToken - Token to read the issuer url from.
 * @returns The issuer url if available.
 */
export function issuerUrl(decodedToken: DecodedJWT): string | undefined {
  return readPropertyWithWarn(decodedToken, 'iss');
}

/**
 * Get the user id of a decoded JWT.
 * @param decodedToken - Token to read the user id from.
 * @returns The user id if available.
 */
export function userId(decodedToken: DecodedJWT): string | undefined {
  return readPropertyWithWarn(decodedToken, mappingUserFields.id.keyInJwt);
}

/**
 * Get the user's given name of a decoded JWT.
 * @param decodedToken - Token to read the user id from.
 * @returns The user id if available.
 */
export function userGivenName(decodedToken: DecodedJWT): string | undefined {
  if (mappingUserFields.givenName) {
    return readPropertyWithWarn(
      decodedToken,
      mappingUserFields.givenName.keyInJwt
    );
  }
}

/**
 * Get the user's family name of a decoded JWT.
 * @param decodedToken - Token to read the user id from.
 * @returns The user id if available.
 */
export function userFamilyName(decodedToken: DecodedJWT): string | undefined {
  if (mappingUserFields && mappingUserFields.familyName) {
    return readPropertyWithWarn(
      decodedToken,
      mappingUserFields!.familyName.keyInJwt
    );
  }
}

/**
 * Get the user name of a decoded JWT.
 * @param decodedToken - Token to read the user id from.
 * @returns The user id if available.
 */
export function userName(decodedToken: DecodedJWT): string | undefined {
  return readPropertyWithWarn(
    decodedToken,
    mappingUserFields.userName.keyInJwt
  );
}

/**
 * Get the user's email of a decoded JWT.
 * @param decodedToken - Token to read the user id from.
 * @returns The user id if available.
 */
export function userEmail(decodedToken: DecodedJWT): string | undefined {
  if (mappingUserFields && mappingUserFields.email) {
    return readPropertyWithWarn(decodedToken, mappingUserFields.email.keyInJwt);
  }
}

/**
 * Get the user's scopes of a decoded JWT.
 * @param decodedToken - Token to read the user id from.
 * @returns The user id if available.
 */
export function userScopes(decodedToken: DecodedJWT): Scope[] | [] {
  if (!(decodedToken.scope instanceof Array && decodedToken.scope.length)) {
    return [];
  }
  return decodedToken.scope
    .map(s => (s.includes('.') ? s.substr(s.indexOf('.') + 1, s.length) : s))
    .map(s => ({ name: s }));
}

/**
 * Get the tenant id of a decoded JWT.
 * @param decodedToken - Token to read the tenant id from.
 * @returns The tenant id if available.
 */
export function tenantId(decodedToken: DecodedJWT): string | undefined {
  return readPropertyWithWarn(decodedToken, mappingTenantFields.id.keyInJwt);
}

/**
 * Extracts the custom attributes in the JWT
 * @param decodedToken - Token to read the custom attributes
 * @returns custom attributes added by the xsuaa to the issued JWT.
 */
export function customAttributes(
  decodedToken: DecodedJWT
): Map<string, string[]> {
  if (decodedToken[mappingUserFields.customAttributes.keyInJwt]) {
    return readPropertyWithWarn(
      decodedToken,
      mappingUserFields.customAttributes.keyInJwt
    ) as Map<string, string[]>;
  }
  return new Map<string, string[]>();
}

/**
 * Get the tenant name of a decoded JWT.
 * @param decodedToken - Token to read the tenant id from.
 * @returns The tenant id if available.
 */
export function tenantName(decodedToken: DecodedJWT): string | undefined {
  const extAttr = readPropertyWithWarn(decodedToken, 'ext_attr');
  if (extAttr) {
    return readPropertyWithWarn(extAttr, 'zdn');
  }
  return undefined;
}

/**
 * Retrieve the audiences of a decoded JWT based on the audiences and scopes in the token.
 * @param decodedToken - Token to retrieve the audiences from.
 * @returns A set of audiences.
 */
// Comments taken from the Java SDK implementation
// Currently, scopes containing dots are allowed.
// Since the UAA builds audiences by taking the substring of scopes up to the last dot,
// Scopes with dots will lead to an incorrect audience which is worked around here.
// If a JWT contains no audience, infer audiences based on the scope names in the JWT.
// This is currently necessary as the UAA does not correctly fill the audience in the user token flow.
export function audiences(decodedToken: DecodedJWT): Set<string> {
  if (audiencesFromAud(decodedToken).length) {
    return new Set(audiencesFromAud(decodedToken));
  }
  return new Set(audiencesFromScope(decodedToken));
}

function audiencesFromAud(decodedToken: DecodedJWT): string[] {
  if (!(decodedToken.aud instanceof Array && decodedToken.aud.length)) {
    return [];
  }
  return decodedToken.aud.map(aud =>
    aud.includes('.') ? aud.substr(0, aud.indexOf('.')) : aud
  );
}

function audiencesFromScope(decodedToken: DecodedJWT): string[] {
  if (!decodedToken.scope) {
    return [];
  }

  const scopes =
    decodedToken.scope instanceof Array
      ? decodedToken.scope
      : [decodedToken.scope];
  return scopes.reduce((aud, scope) => {
    if (scope.includes('.')) {
      return [...aud, scope.substr(0, scope.indexOf('.'))];
    }
    return aud;
  }, []);
}

/**
 * Wraps the access token in header's authorization.
 * @param token - Token to attach in request header
 * @returns The request header that holds the access token
 */
export function wrapJwtInHeader(token: string): AxiosRequestConfig {
  return { headers: { Authorization: 'Bearer ' + token } };
}

function readPropertyWithWarn(decodedJwt: DecodedJWT, property: string): any {
  if (!decodedJwt[property]) {
    logger.warn(
      `WarningJWT: The provided JWT does not include "${property}" property.`
    );
  }

  return decodedJwt[property];
}

/**
 * Fetches the URL from the JWT header which exposes the verification key for that JWT.
 * @param token - Undecoded JWT as a string
 * @returns The value of jku property of the JWT header
 */
function getVerificationKeyURL(token: string): string {
  const decodedJwt = decodeJwtComplete(token);
  if (!decodedJwt.header.jku) {
    throw new Error(
      'Field jku containing the URL for JWT verification not part of header.'
    );
  }
  return decodedJwt.header.jku;
}

/**
 * Interface to represent the registered claims of a JWT.
 */
export type RegisteredJWTClaims = RegisteredJWTClaimsBasic &
  RegisteredJWTClaimsUser &
  RegisteredJWTClaimsTenant;

/**
 * Interface to represent the basic properties like issuer, audience etc.
 */
export interface RegisteredJWTClaimsBasic {
  iss?: string;
  exp?: number;
  sub?: string;
  aud?: string[];
  nbf?: string;
  iat?: number;
  jti?: string;
}

/**
 * Interface to represent the basic properties of a jwt header
 */
export interface JWTHeader {
  alg: string;
  typ: string;
  jku?: string;
}

/**
 * @Deprecated Use [[JWTPayload]] if you want to represent the decoded JWT payload or [[CompleteDecodedJWT]] for the full decoded object.
 * Interface to represent the payload of a JWT.
 */
export interface DecodedJWT extends RegisteredJWTClaims {
  [otherKey: string]: any;
}

/**
 * Interface to represent the payload of a JWT.
 */
export interface JWTPayload extends RegisteredJWTClaims {
  [otherKey: string]: any;
}

/**
 * Interface to represent header and  payload of a JWT.
 */
export interface CompleteDecodedJWT extends RegisteredJWTClaims {
  header: JWTHeader;
  payload: JWTPayload;
  signature: string;
}

export type JwtKeyMapping<TypescriptKeys, JwtKeys> = {
  [key in keyof TypescriptKeys]: {
    keyInJwt: keyof JwtKeys;
    extractorFunction: (decodedJWT: DecodedJWT) => any;
  };
};

/**
 * Checks if a given key is in the decoded JWT. If not an error is raised
 * @param key - The key of the representation in typescript
 * @param mapping - The mapping between the typescript keys and the JWT key
 * @param decodedJWT - Decoded token on which the check is done
 * @exception Error is thrown if the key is not present.
 */
export function checkMandatoryValue<TypeScriptKeys, JwtKeys>(
  key: keyof TypeScriptKeys,
  mapping: JwtKeyMapping<TypeScriptKeys, JwtKeys>,
  decodedJWT: DecodedJWT
): void {
  const value = mapping[key].extractorFunction(decodedJWT);
  if (!value) {
    throw new Error(
      `Field ${mapping[key].keyInJwt} not provided in decoded jwt: ${decodedJWT}`
    );
  }
}
