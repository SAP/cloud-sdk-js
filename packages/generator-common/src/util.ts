import { codeBlock } from '@sap-cloud-sdk/util';

/**
 * @returns A copyright header
 * @internal
 */
export function getCopyrightHeader(): string {
  return codeBlock`
/*
 * Copyright (c) ${new Date().getFullYear()} SAP SE or an SAP affiliate company. All rights reserved.
 *
 * This is a generated file powered by the SAP Cloud SDK for JavaScript.
 */
 `;
}
