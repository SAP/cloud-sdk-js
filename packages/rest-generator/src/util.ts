import { promises } from 'fs';
import { join } from 'path';
import { codeBlock } from '@sap-cloud-sdk/util';
const { writeFile, mkdir, rmdir } = promises;

export function createDirectory(directoryPath: string): Promise<void> {
  return mkdir(directoryPath, { recursive: true });
}

export function createFile(
  directoryPath: string,
  fileName: string,
  content: string,
  overwrite: boolean
): Promise<void> {
  return writeFile(join(directoryPath, fileName), wrapContent(content), {
    encoding: 'utf8',
    flag: overwrite ? 'w' : 'wx'
  });
}

function wrapContent(content: string): string {
  return (
    codeBlock`
/*
 * Copyright (c) ${new Date().getFullYear()} SAP SE or an SAP affiliate company. All rights reserved.
 *
 * This is a generated file powered by the SAP Cloud SDK for JavaScript.
 */
${content}
` + '\n'
  );
}

export function cleanDirectory(directoryPath: string): Promise<void> {
  return rmdir(directoryPath, { recursive: true });
}
