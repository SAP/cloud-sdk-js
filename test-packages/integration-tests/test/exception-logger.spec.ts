import { resolve } from 'path';
import execa from 'execa';
import { createLogger, kibana, local } from '@sap-cloud-sdk/util';
import { Logger } from 'winston';

describe('exception logger', () => {
  let logger: Logger;

  beforeEach(() => {
    if (logger) {
      logger.close();
    }
  });

  it('should log exception with local format if they fly in development mode', async () => {
    process.env.NODE_ENV = 'development';
    await expect(
      execa('ts-node', [
        resolve(__dirname, 'throw-exception-with-logger-script.ts')
      ])
    ).rejects.toThrowError(/Test Exception Logger\n\s*at Object/);
    delete process.env.NODE_ENV;
  }, 15000);

  it('should log exception with kibana format if they fly in production mode', async () => {
    process.env.NODE_ENV = 'production';
    await expect(
      execa('ts-node', [
        resolve(__dirname, 'throw-exception-with-logger-script.ts')
      ])
    ).rejects.toThrowError(/Test Exception Logger\\n\s*at Object/);
    delete process.env.NODE_ENV;
  }, 15000);

  it('should not log the stack multiple times', async () => {
    try {
      await execa('ts-node', [
        resolve(__dirname, 'throw-exception-with-logger-script.ts')
      ]);
    } catch (err) {
      expect(err.message.match(/Test Exception Logger/g).length).toBe(1);
    }
  }, 15000);

  it('logs stack trace of exception locally', () => {
    const tranformerSpy = jest.spyOn(local, 'transform');
    logger = createLogger('stack-logger-local');
    logger.error(new Error('My Error'));

    expect(tranformerSpy).toReturnWith(
      expect.objectContaining({
        message: expect.stringMatching(/My Error\n\s+at\sObject/)
      })
    );
  });

  it('logs stack trace of exception on CF', () => {
    const tranformerSpy = jest.spyOn(kibana, 'transform');
    logger = createLogger('stack-logger-cf');
    logger.format = kibana;
    logger.error(new Error('My Error'));

    expect(tranformerSpy).toReturnWith(
      expect.objectContaining({
        msg: expect.stringMatching(/My Error\n\s+at\sObject/)
      })
    );
  });

  it('logs message with padding if string is given', () => {
    const tranformerSpy = jest.spyOn(local, 'transform');
    logger = createLogger('stack-logger-string');
    logger.error('My Error.');
    expect(tranformerSpy).toReturnWith(
      expect.objectContaining({ message: '   My Error.' })
    );
  });

  it('does not log stack on lower log levels', () => {
    const tranformerSpy = jest.spyOn(local, 'transform');
    logger = createLogger('stack-logger-string');
    logger.info(new Error('My Error'));
    expect(tranformerSpy).toReturnWith(
      expect.objectContaining({ message: '   My Error.' })
    );
  });
});
