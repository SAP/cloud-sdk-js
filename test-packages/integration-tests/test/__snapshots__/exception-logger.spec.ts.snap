// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exception logger should not log the stack twice 1`] = `
"Command failed: ts-node /Users/d068544/WebstormProjects/cloud-sdk-second-checkout/test-packages/integration-tests/test/throw-exception-with-logger-script.ts

[90m[2021-02-17T16:23:39.116Z][39m [32mINFO[39m     [34m(throw-exception-with-logger)[39m: Some log statement.
[90m[2021-02-17T16:23:39.159Z][39m [7m[31mERROR[39m[27m    Error: Test Exception Logger
    at Object.<anonymous> (/Users/d068544/WebstormProjects/cloud-sdk-second-checkout/test-packages/integration-tests/test/throw-exception-with-logger-script.ts:5:7)
    at Module._compile (internal/modules/cjs/loader.js:1123:30)
    at Module.m._compile (/Users/d068544/WebstormProjects/cloud-sdk-second-checkout/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1143:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/d068544/WebstormProjects/cloud-sdk-second-checkout/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (internal/modules/cjs/loader.js:972:32)
    at Function.Module._load (internal/modules/cjs/loader.js:872:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at main (/Users/d068544/WebstormProjects/cloud-sdk-second-checkout/node_modules/ts-node/src/bin.ts:198:14)
    at Object.<anonymous> (/Users/d068544/WebstormProjects/cloud-sdk-second-checkout/node_modules/ts-node/src/bin.ts:288:3)
"
`;
